src/main/java/com/google/sps/data/GridCell.java

```diff
@@ -0,0 +1,22 @@
+package com.google.sps.data;
+
+public class GridCell {
+  private int numPoints;
+  public double averageAQI;
+
+  public GridCell() {
+    this.numPoints = 0;
+    this.averageAQI = 0;
+  }
+  /**
+   * Update the averageAQI, taking into account the number of points
+   * already added to avoid creating bias towards the new point
+   */
+  public void addPoint(AQDataPoint point) {
+    if (point.aqi == -1) {
+        return;
+    }
+    averageAQI = (averageAQI * numPoints + point.aqi) / (numPoints + 1);
+    numPoints ++;
```

Comment:
I did not realise that they were sorted by hour...

It could work with multiple data sources, but there would be an extra step of sorting all of the data points by time first, which would get the time weighted average, but I'm not sure if the extra complexity is worthwhile right now...

Reply:
It's an implementation detail of the NSWGovAdapter, which should not really be relied on. In order to reliably handle time, you'd need to add a time component to the AQDataPoint, and correctly annotate it.

This was more a random food for thought, you choose to do a unbiased weighting of all data points landing within the same grid. Why that? Instead of ... dropping all duplicates? Keeping the last value? If it was just arbitrary, that's ok too, but it's good to think about why your processing something in a particular way, and validate any assumptions that makes about the data.

For example, in this particular case, taking an unbiased average will give a less "accurate" view of the current AQ, because of what the Adapter is passing back currently. Keeping the last of any conflicting values will probably be the most accurate. For now.

Reply:
With the larger grid sizes, it didn't make sense do only keep one value, because if they were at opposite corners then they might show some trend/change that happens in that grid cell, and defining duplicates is hard with coords because of floating points. I think it would make more sense to only keep one value with the static smaller grid cell sizes though, but the averaging was only an interim approach to get the data into the grid in a somewhat meaningful way.

Reply:
Again; just food for thought. Probably worth re-visiting once the grid method is updated, and the secondary data source is added.